### PR TITLE
RD-5176 - don't retry to send after timeout to edge

### PR DESCRIPTION
**Pull request by saartochner**

### CODE REVIEW

File: src/lumigo_tracer/lumigo_utils.py [hunk 1]
```diff
@@ -106,6 +106,16 @@ def get_region() -> str:
     pass
 
 
+class InternalState:
+    timeout_on_connection = False
+    internal_error_already_logged = False
+
+    @staticmethod
+    def reset():
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:114]: You call `reset` only in conftest. Don't you need to call it also at the beginning of the tracer execution? Won't it leak when there's a warm start?

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice :)

File: src/lumigo_tracer/lumigo_utils.py (hunk 1 above)
Comment (saartochner) [reply src/lumigo_tracer/lumigo_utils.py:114]: It should leak. This is a global state that we should have :)

Comment (saartochner) [reply src/lumigo_tracer/lumigo_utils.py:114]: E.g. - when there is a timeout to the edge, don't waste more time on it (even in future invocations)

File: src/lumigo_tracer/lumigo_utils.py [hunk 2]
```diff
@@ -107,14 +109,25 @@ def get_region() -> str:
...
-        InternalState.timeout_on_connection = False
+        InternalState.timeout_on_connection = None
         InternalState.internal_error_already_logged = False
 
+    @staticmethod
+    def mark_timeout_to_edge():
+        InternalState.timeout_on_connection = datetime.datetime.now()
+
+    @staticmethod
+    def should_report_to_edge():
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:125]:
```suggestion
    def should_report_to_edge() -> bool:
```

File: src/lumigo_tracer/lumigo_utils.py [hunk 3]
```diff
@@ -34,6 +35,7 @@
 HTTPS_PREFIX = "https://"
 LOG_FORMAT = "#LUMIGO# - %(asctime)s - %(levelname)s - %(message)s"
 SECONDS_TO_TIMEOUT = 0.5
+COOLDOWN_AFTER_TIMEOUT_DURATION = 10
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:38]: IMO this is easier to understand
```suggestion
COOLDOWN_AFTER_TIMEOUT_DURATION = datetime.timedelta(seconds=10)
```

**Review by nirLumigo — APPROVED** [review]

File: src/lumigo_tracer/lumigo_utils.py [hunk 4]
```diff
@@ -753,3 +770,17 @@ def is_provision_concurrency_initialization() -> bool:
 def get_stacktrace(exception: Exception) -> str:
     original_traceback = traceback.format_tb(exception.__traceback__)
     return "".join(filter(lambda line: STACKTRACE_LINE_TO_DROP not in line, original_traceback))
+
+
+try:
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:775]: moved from above to reuse `establish_connection`

**Review by nirLumigo — APPROVED** [review]

**Pull request merged**